forem/forem
# Cache return value of `FeatureFlag.enabled?`

## What type of PR is this? (check all applicable)


## Description

The [top query on DEV by time spent](https://app.datadoghq.com/apm/service/practical_developer-postgres/postgres.query?env=production&sort=duration%2Cdesc&topGraphs=latency%3Alatency%2Cerrors%3Acount%2Chits%3Acount&start=1630610127121&end=1630696527121&paused=false) is the query to get feature flags:

<img width="658" alt="" src="https://user-images.githubusercontent.com/108205/132055741-51bcc500-94f0-4d88-b485-f3953f67b383.png">

There is an additional query a little further down that checks flipper gates:

<img width="455" alt="image" src="https://user-images.githubusercontent.com/108205/132056714-1fd1d290-e86d-4968-8c5c-c704f1f05583.png">

This PR adds a `CACHE_FEATURE_FLAGS_SECONDS` environment variable that allows caching both of these for the specified period of time. The cache duration is designed to be tunable without a deploy and will likely be 5 seconds to start — long enough to reduce DB load but not so long that toggling a feature flag would cause us to have to wait.

We use an in-memory store for 3 reasons:

1. The number of feature flags we use is pretty low, so it won't bloat memory by any meaningful amount
2. Using `Rails.cache` would simply transfer load from Postgres to Redis
   - To be fair, Redis could almost certainly handle this extra load just fine. I've pushed 15k queries/sec through Redis (well, AWS ElastiCache) in production before. However, we also often see nontrivial query times on our Redis cache instance, so I don't want exacerbate that.
   - Using an in-memory cache also avoids application overhead involved in syscalls (I/O, memory allocation, etc) when unmarshaling cache data coming in over the wire. The objects don't need to be unmarshaled if they're already in memory.
3. It's okay if each instance of the application fetches fresh data once every 5 seconds or so. It's still a far cry better than hundreds or thousands of times per second.

## QA Instructions, Screenshots, Recordings

In a Rails console:

- Run `FeatureFlag.enabled? :foo` several times
  - The first time, you should see `Cache read` and`Cache generate` along with an ActiveRecord query
  - Subsequent invocations should show `Cache read` and `Cache fetch_hit` while omitting the ActiveRecord query

## Added/updated tests?

- [x] No, and this is why: Configuration change

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Admin Guide](https://admin.forem.com/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
      or in a [forem.dev](http://forem.dev) post
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

- Set `CACHE_FEATURE_FLAGS_SECONDS=5` on DEV
- Check the hit rate against `flipper_*` tables on [this Datadog page](https://app.datadoghq.com/apm/service/practical_developer-postgres/postgres.query?env=production&sort=duration%2Cdesc&topGraphs=latency%3Alatency%2Cerrors%3Acount%2Chits%3Acount&start=1630611223218&end=1630697623218&paused=false). It should be less than 1/sec.
